### PR TITLE
Add calibration statistics

### LINKED ISSUE #298: Calibration measures without binning, and associated p-values.
MAPIE computes calibration metrics in binary settingd such as Expected Calibration Error (ECE). However, these metrics rely on a an arbitrary binning scheme, that make them unstable. Recently, some researchers at Facebook provided alternatives to measure calibration. The benefits of these new methods are : 
- They do not rely on any binning scheme
- They provide visualisation easy to interpret that facilitates the investigation of miscalibration
- They provide statistical tests backed by strong theoretical guarantees, that allows to test the null hypothesis "the predictions are calibrated"

Attached is a summary of these two articles, in french :

[Arrieta et al. 2022](https://arxiv.org/abs/2205.09680)
Capture d’écran 2023-04-14 à 21 46 02

[Tygert et al. 2022](https://arxiv.org/abs/2202.00100)
Capture d’écran 2023-04-14 à 21 46 16

These papers are straightforward to implement, as exemplified in this notebook : 

[p_values_for_calibration.pdf](https://github.com/scikit-learn-contrib/MAPIE/files/11236580/p_values_for_calibration.pdf)

Since MAPIE uses binary calibration as a core component (e.g. in top-label calibration), I think I could propose quite rapidly a good quality pull request to MAPIE providing these new calibration measures and associated tests.

Do you consider the contribution would be in line with MAPIE roadmap ? Do you have guidelines on where to insert such pieces of code ?

Thanks in advance for your response !

### PR DESCRIPTION
**Pull request by gmartinonQM**
# Description

Enrich metrics with binary calibration statistical tests taken from this paper : [Arrieta et al. 2022](https://arxiv.org/abs/2205.09680)

Fixes #298

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

All function are exemplified and tested, results from the paper are reproduced to machine precision, an illustrative notebook on a toy example confirms the intuition

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully : `make doc`

### CODE REVIEW

**Review by thibaultcordier — COMMENTED** [review]
Your additions are a must-have for mapie and the quantification of calibration quality. I took the liberty of adding some comments to better understand what we're trying to calculate and what results the user should expect. 

By the way, calculating statistics is very similar to statistical tests as we're used to seeing them. Do you think adapting your statistics to tests (with calculation of p-values) is a good idea to make contributions user-friendly or not?

Ref: https://arxiv.org/abs/2202.00100 (in biblio of https://arxiv.org/abs/2205.09680).

File: mapie/metrics.py [hunk 1]
```diff
@@ -737,3 +738,305 @@ def hsic(
...
+    """
+    s = cast(NDArray, column_or_1d(s))
+    n = len(s)
+    length_scale = np.sqrt(np.sum(s * (1 - s)))/n
+    return length_scale
+
+
+def kolmogorov_smirnov_statistic(y_true: NDArray, y_score: NDArray) -> float:
+    """
+    Compute Kolmogorov-smirnov's statistic for calibration test.
```
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/metrics.py:921]: Done

File: mapie/metrics.py [hunk 2]
```diff
@@ -737,3 +738,305 @@ def hsic(
...
+    y_score = cast(NDArray, column_or_1d(y_score))
+    cum_diff = cumulative_differences(y_true, y_score)
+    sigma = length_scale(y_score)
+    ks_stat = np.max(np.abs(cum_diff)) / sigma
+    return ks_stat
+
+
+def kuiper_statistic(y_true: NDArray, y_score: NDArray) -> float:
+    """
+    Compute Kuiper's statistic for calibration test.
```
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/metrics.py:961]: Done

File: mapie/metrics.py [hunk 3]
```diff
@@ -737,3 +738,305 @@ def hsic(
...
+    y_score = cast(NDArray, column_or_1d(y_score))
+    cum_diff = cumulative_differences(y_true, y_score)
+    sigma = length_scale(y_score)
+    ku_stat = (np.max(cum_diff) - np.min(cum_diff)) / sigma
+    return ku_stat
+
+
+def spiegelhalter_statistic(y_true: NDArray, y_score: NDArray) -> float:
+    """
+    Compute Spiegelhalter's statistic for calibration test.
```
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/metrics.py:1001]: Done

**Review by thibaultcordier — APPROVED** [review]
Excellent PR, the tests are complete and the functions have been added to the API. It's ok for me. I suggest you have an example in the gallery in another PR.

File: mapie/metrics.py [hunk 4]
```diff
@@ -1,7 +1,9 @@
-from typing import Optional, cast, Union
+from typing import Optional, cast, Union, Tuple
 
+import scipy
 import numpy as np
 from sklearn.utils.validation import check_array, column_or_1d
+from sklearn.utils import check_random_state
```
Comment (LacombeLouis) [inline mapie/metrics.py:6]: Please add the imports in alphabetical order :)

File: mapie/metrics.py [hunk 5]
```diff
@@ -737,3 +740,578 @@ def hsic(
     coef_hsic = np.sqrt(np.matrix.trace(hsic_mat, axis1=1, axis2=2))
 
     return coef_hsic
+
+
+def jitter(
```
Comment (LacombeLouis) [inline mapie/metrics.py:745]: Not sure about the name of this function, would `add_jitter` not be more explicit?

File: mapie/metrics.py [hunk 6]
```diff
@@ -737,3 +740,578 @@ def hsic(
...
+    >>> from mapie.metrics import sort_xy_by_y
+    >>> x = np.array([1, 2, 3, 4, 5])
+    >>> y = np.array([5, 4, 3, 1, 2])
+    >>> x_sorted, y_sorted = sort_xy_by_y(x, y)
+    >>> print(x_sorted)
+    [4 5 3 2 1]
+    >>> print(y_sorted)
+    [1 2 3 4 5]
+    """
+    x = cast(NDArray, column_or_1d(x))
```
Comment (LacombeLouis) [inline mapie/metrics.py:815]: Why do you cast to `NDArray` if that's already the type of the variable as an input?

File: mapie/metrics.py [hunk 7]
```diff
@@ -737,3 +740,578 @@ def hsic(
...
+    Example
+    -------
+    >>> import numpy as np
+    >>> from mapie.metrics import kolmogorov_smirnov_statistic
+    >>> y_true = np.array([0, 1, 0, 1, 0])
+    >>> y_score = np.array([0.1, 0.9, 0.21, 0.9, 0.5])
+    >>> print(np.round(kolmogorov_smirnov_statistic(y_true, y_score), 3))
+    0.978
+    """
+    y_true = cast(NDArray, column_or_1d(y_true))
```
Comment (LacombeLouis) [inline mapie/metrics.py:964]: Here also, why the casting?

File: mapie/metrics.py [hunk 8]
```diff
@@ -737,3 +740,578 @@ def hsic(
...
+    -------
+    >>> import pandas as pd
+    >>> from mapie.metrics import kuiper_p_value
+    >>> y_true = np.array([1, 0, 1, 0, 1, 0])
+    >>> y_score = np.array([0.8, 0.3, 0.5, 0.5, 0.7, 0.1])
+    >>> ku_p_value = kuiper_p_value(y_true, y_score)
+    >>> print(np.round(ku_p_value, 4))
+    0.9684
+    """
+    ku_stat = kuiper_statistic(y_true, y_score)
```
Comment (LacombeLouis) [inline mapie/metrics.py:1220]: You do not check for `column_or_1d`?

File: mapie/metrics.py [hunk 9]
```diff
@@ -737,3 +740,578 @@ def hsic(
     coef_hsic = np.sqrt(np.matrix.trace(hsic_mat, axis1=1, axis2=2))
 
     return coef_hsic
+
+
+def jitter(
+    x: NDArray,
+    noise_amplitude: float = 1e-8,
```
Comment (LacombeLouis) [inline mapie/metrics.py:747]: Can the `noise_amplitude` be negative?

**Review by LacombeLouis — COMMENTED** [review]
Thank you Grégoire for this PR! This will bring crucial additional metrics to better understand if the algorithm has calibrated score outputs. 

On that note, it would be interesting if you had some examples and if you could bring some clarifications as to in which context to use these metrics.

File: mapie/metrics.py (hunk 5 above)
Comment (gmartinonQM) [reply mapie/metrics.py:745]: Done

File: mapie/metrics.py (hunk 6 above)
Comment (gmartinonQM) [reply mapie/metrics.py:815]: Good point ! Just copy-pasted from `hsic` function...
remove the cast : done

File: mapie/metrics.py (hunk 7 above)
Comment (gmartinonQM) [reply mapie/metrics.py:964]: Done

File: mapie/metrics.py (hunk 8 above)
Comment (gmartinonQM) [reply mapie/metrics.py:1220]: The check is already done in `kuiper_statistic`. I wondered where to put the check : 
- in *_statistic functions
- or in *_p_value functions
I choose the first option, but both are valid, have no preference about it. What do you think ?

File: mapie/metrics.py (hunk 9 above)
Comment (gmartinonQM) [reply mapie/metrics.py:747]: Yes

File: mapie/metrics.py (hunk 4 above)
Comment (gmartinonQM) [reply mapie/metrics.py:6]: Done

File: mapie/tests/test_metrics.py [hunk 10]
```diff
@@ -21,7 +21,14 @@
                            regression_mean_width_score,
                            regression_ssc,
                            regression_ssc_score,
-                           top_label_ece)
+                           top_label_ece,
```
[the remark this answers is not in the document]
Comment (gmartinonQM) [reply mapie/tests/test_metrics.py:24]: Done

File: doc/theoretical_description_calibration.rst [hunk 11]
```diff
@@ -72,16 +72,115 @@ The ECE is the combination of these two metrics combined together.
...
 In the top-label calibration, we only calculate the ECE for the top-label class. Hence, per top-label class, we condition the calculation
 of the accuracy and confidence based on the top label and take the average ECE for each top-label.
 
+3. Statistical tests for calibration
+------------------------------------
+
+**Kolmogorov-Smirnov test**
+
+Kolmogorov-Smirnov test was derived in [2,3,4]. The idea is to consider the cumulative differences between sorted scores :math:`s_i`
+and their corresponding labels :math:`y_i` and to compare its porperties to that of a standard Brownian motion. Let us consider the
```
Comment (thibaultcordier) [inline doc/theoretical_description_calibration.rst:86]:
```suggestion
and their corresponding labels :math:`y_i` and to compare its properties to that of a standard Brownian motion. Let us consider the
```

File: doc/api.rst [hunk 12]
```diff
@@ -49,19 +49,31 @@ Metrics
...
+   metrics.regression_mean_width_score
    metrics.regression_ssc
    metrics.regression_ssc_score
-   metrics.classification_ssc
-   metrics.classification_ssc_score
-   metrics.hsic
+   metrics.sort_xy_by_y
+   metrics.spiegelhalter_p_value
+   metrics.spiegelhalter_statistic
+   metrics.top_label_ece
```
Comment (thibaultcordier) [inline doc/api.rst:76]: On the user side, I don't think users need access to the following functions: add_jitter, cumulative_differences, length_scale, sort_xy_by_y. I see them more as utilities for other metrics.

Do you think it's necessary to present them in the API?

**Review by thibaultcordier — COMMENTED** [review]
I left you a comment about the API. I'm looking for a solution to generate the documentation without any problem.

File: doc/theoretical_description_calibration.rst [hunk 13]
```diff
@@ -72,16 +72,115 @@ The ECE is the combination of these two metrics combined together.
...
+cumulative differences on sorted scores: 
+
+.. math::
+    \C_k = \frac{1}{N}\sum_{i=1}^k (s_i - y_i)
```
Comment (thibaultcordier) [inline doc/theoretical_description_calibration.rst:90]:
```suggestion
    C_k = \frac{1}{N} \sum_{i=1}^k (s_i - y_i)
```

**Comment by thibaultcordier** [comment]
@gmartinonQM, the problem of compiling documentation is now solved. I suggest you merge your branch with the master branch as follows (in your local branch):
```bash
git merge master
```
Consult https://git-scm.com/docs/git-merge if you have any difficulties, or ask me.

File: doc/theoretical_description_calibration.rst (hunk 11 above)
Comment (gmartinonQM) [reply doc/theoretical_description_calibration.rst:86]: Done

File: doc/api.rst (hunk 12 above)
Comment (gmartinonQM) [reply doc/api.rst:76]: Agreed for `add_jitter` and `sort_xy_by_y` but I would leave `cumulative_differences` and `length_scale` as they are used in the examples.

File: doc/theoretical_description_calibration.rst [hunk 14]
```diff
@@ -72,16 +72,115 @@ The ECE is the combination of these two metrics combined together.
...
+converges to the maximum absolute value of a standard Brownian motion over the unit interval :math:`[0, 1]`. [3, 4] also provide closed-form 
+formulas for the cumulative distribution function (CDF) of the maximum absolute value of such a standard Brownian motion.
+So we state the p-value associated to the statistical test of well calibration as:
+
+.. math::
+   p = 1 - CDF(G)
+
+**Kuiper test**
+
+Kuiper test was derived in [2, 3, 4] and is very similar to Kolmogorov-Smirnov. This time, the statistic is defined as
```
Comment (thibaultcordier) [inline doc/theoretical_description_calibration.rst:112]:
```suggestion
Kuiper test was derived in [2, 3, 4] and is very similar to Kolmogorov-Smirnov. This time, the statistic is defined as:
```

**Review by thibaultcordier — APPROVED** [review]
Excellent PR with useful documentation. I just have a few suggestions to add before the merge.

**Review by LacombeLouis — APPROVED** [review]
